jbpm-wb-case-mgmt/jbpm-wb-case-mgmt-client/src/main/java/org/jbpm/workbench/cm/client/roles/CaseRolesPresenter.java

```diff
@@ -16,201 +16,223 @@
+                }
+        ).getCaseDefinition(serverTemplateId, containerId, cis.getCaseDefinitionId());
+    }
 
-                    final Set<String> roles = getRolesAvailableForAssignment(cis, cds);
-                    if (roles.isEmpty()) {
-                        return;
-                    }
+    private boolean hasAssignment(CaseRoleAssignmentSummary caseRoleAssignmentSummary) {
+        return caseRoleAssignmentSummary.getUsers().size() + caseRoleAssignmentSummary.getGroups().size() > 0;
 
-                    view.enableNewRoleAssignments();
+    }
 
-                    view.setUserAddCommand(() ->
-                            newRoleAssignmentView.show( roles, () ->
-                                    addAssignment(newRoleAssignmentView.getUserName(),
-                                            newRoleAssignmentView.getGroupName(),
-                                            newRoleAssignmentView.getRoleName())));
-                }
-        ).getCaseDefinition(serverTemplateId, containerId, cis.getCaseDefinitionId());
+    public void filterElements() {
+        String filterBy = view.getFilterValue();
+
+        if (filterBy.equals(CaseRolesAssignmentFilterBy.ASSIGNED.getLabel())) {
+            view.setRolesAssignmentList(
+                    allUnfilteredElements.stream()
+                            .filter(caseRoleAssignmentSummary -> hasAssignment(caseRoleAssignmentSummary))
+                            .collect(toList()));
+        } else if (filterBy.equals(CaseRolesAssignmentFilterBy.UNASSIGNED.getLabel())) {
+            view.setRolesAssignmentList(
+                    allUnfilteredElements.stream()
+                            .filter(caseRoleAssignmentSummary -> !hasAssignment(caseRoleAssignmentSummary))
+                            .collect(toList()));
+        } else {
+            view.setRolesAssignmentList(allUnfilteredElements);
+        }
     }
 
-    protected Set<String> getRolesAvailableForAssignment(final CaseInstanceSummary cis, final CaseDefinitionSummary cds) {
-        return cds.getRoles().keySet().stream().filter(
-                role -> {
-                    if ("owner".equals(role)) {
-                        return false;
-                    }
-                    final Integer roleCardinality = cds.getRoles().get(role);
-                    if (roleCardinality == -1) {
-                        return true;
-                    }
-                    final Integer roleInstanceCardinality = cis.getRoleAssignments().stream().filter(ra -> role.equals(ra.getName())).findFirst().map(ra -> ra.getGroups().size() + ra.getUsers().size()).orElse(0);
-                    return roleInstanceCardinality < roleCardinality;
-                }
-        ).collect(Collectors.toSet());
+    public void editAction(CaseRoleAssignmentSummary caseRoleAssignmentSummary) {
+        CaseRoleAssignmentSummary previousCaseRoleAssignmentSummary= getRoleAssignmentByName(caseRoleAssignmentSummary.getName());
```

Comment:
It looks to me that the only piece of data which connects _editAction_, _assignToRole_ and _storeRoleAssignments_ methods is the role name in  _CaseRoleAssignmentSummary_ model that is passed to _editAction_.
Maybe it would be more readable if we could make this connection clear and strip away methods from unnecessary computing of data which they don't use.
I'm thinking something along the lines of:
- remove _getRoleAssignmentByName_ call in _editAction_
- change _assignToRole_ signature so it doesn't take any arguments (the ones it takes now it doesn't use but just forwards)
- remove _oldUserAssignments_ and _oldGroupsAssignments_ lists from _storeRoleAssignments_ arguments and get them from that **cras** reference before you start changing the object it points to.